mithril-common/src/cardano_transactions_preloader.rs

```diff
@@ -0,0 +1,212 @@
+//! # Cardano Transaction Preloader
+//!
+//! This module provides a preload mechanism for Cardano Transaction signed entity, allowing
+//! to compute in advance the Transactions & Block Range Root to be signed.
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use slog::{debug, trace, Logger};
+
+use crate::chain_observer::ChainObserver;
+use crate::entities::{CardanoTransactionsSigningConfig, SignedEntityTypeDiscriminants};
+use crate::signable_builder::TransactionsImporter;
+use crate::signed_entity_type_lock::SignedEntityTypeLock;
+use crate::StdResult;
+
+/// Preload mechanism for Cardano Transaction signed entity, allowing
+/// to compute in advance the Transactions & Block Range Root to be signed.
+pub struct CardanoTransactionsPreloader {
+    signed_entity_type_lock: Arc<SignedEntityTypeLock>,
+    importer: Arc<dyn TransactionsImporter>,
+    cardano_transactions_signing_config: CardanoTransactionsSigningConfig,
+    chain_observer: Arc<dyn ChainObserver>,
+    logger: Logger,
+}
+
+impl CardanoTransactionsPreloader {
+    /// Create a new instance of `CardanoTransactionPreloader`.
+    pub fn new(
+        signed_entity_type_lock: Arc<SignedEntityTypeLock>,
+        importer: Arc<dyn TransactionsImporter>,
+        cardano_transactions_signing_config: CardanoTransactionsSigningConfig,
+        chain_observer: Arc<dyn ChainObserver>,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            signed_entity_type_lock,
+            importer,
+            cardano_transactions_signing_config,
+            chain_observer,
+            logger,
+        }
+    }
+
+    /// Preload the Cardano Transactions by running the importer up to the current chain block number.
+    pub async fn preload(&self) -> StdResult<()> {
+        debug!(self.logger, "🔥 Preload Cardano Transactions - Started");
+        trace!(self.logger, "🔥 Locking signed entity type"; "entity_type" => "CardanoTransactions");
```

Comment:
Maybe we can put higher log level?
```suggestion
        info!(self.logger, "🔥 Preload Cardano Transactions - Started");
        debug!(self.logger, "🔥 Locking signed entity type"; "entity_type" => "CardanoTransactions");
```